### ISSUE TITLE
Field energy NaN values for a 2D set-up.

### ISSUE THREAD

**Issue by Anton-Le** [ComputationalRadiationPhysics/picongpu#4201]
While setting up a 2D simulation with a colleague we have encountered the following problem:

The 2D simulation produces `nan` field values for the field energy from the 1st step onward:
```
#step total[Joule] Bx[Joule] By[Joule] Bz[Joule] Ex[Joule] Ey[Joule] Ez[Joule] 
0 0.000000000000000e+00 0 0 0 0 0 0
5 nan nan nan nan nan nan nan
```
if compiled and run using cde7ccdfa156e24b26074904752274935d323d0a and the hemera V100 profile/TPL therein.

If compiled and run using 88fd0117d35d463a2302b8d2852587a48d22c8b5 the field energies appear correct:
```
#step total[Joule] Bx[Joule] By[Joule] Bz[Joule] Ex[Joule] Ey[Joule] Ez[Joule] 
0 0.000000000000000e+00 0 0 0 0 0 0
5 8.583515945519158e-13 3.58032e-13 5.18102e-15 1.05577e-14 3.98794e-14 3.97184e-14 4.04984e-13
```

The `nan` values do not result in a termination of the simulation. The problem appears in the field values and not only in the field energy computation.

The only error provided in `stderr` pertains to the  PSM2 component of OpenMPI (which, if memory serves, is the integration with SLURM and can be treated as a warning rather than an error):
```
A requested component was not found, or was unable to be opened.  This
means that this component is either not installed or is unable to be
used on your system (e.g., sometimes this means that shared libraries
that the component requires are unable to be found/loaded).  Note that
Open MPI stopped checking at the first component that it did not find.

Host:      gv016.cluster
Framework: mtl
Component: psm2
--------------------------------------------------------------------------
[gv001.cluster:152978] 15 more processes have sent help message help-mca-base.txt / find-available:not-valid
[gv001.cluster:152978] Set MCA parameter "orte_base_help_aggregate" to 0 to see all help / error messages
```

**Comment by sbastrakov** [comment]
On the dev system this setup (not each time, but regularly) crashes and does not produce NaNs. Which is arguably better as it may be easier to catch. With debug information and cuda-gdb it gives `Thread 1 "picongpu" received signal CUDA_EXCEPTION_14, Warp Illegal Address.` [here](https://github.com/ComputationalRadiationPhysics/picongpu/blob/365820926e978deeae135c7fd15e62a43353b184/thirdParty/mallocMC/src/include/mallocMC/creationPolicies/Scatter.hpp#L330). Other debugging attemps indicate it crashes inside collisions, however not clear due to collision implementation itself or memory allocation or something else.

**Comment by sbastrakov** [comment]
As a note, in addition to PIConGPU changes since the version that works for this setup, there were also updates to mallocMC and alpaka/cupla. So if we suspect memory, should also look there.

**Comment by sbastrakov** [comment]
Output of `cuda-memcheck` when compiled with same flags is:
```
========= Invalid __global__ read of size 4
=========     at 0x000052c0 in /opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/targets/x86_64-linux/include/device_atomic_functions.hpp:157:_ZN6alpaka16uniform_cuda_hip6detail20uniformCudaHipKernelINS_9ApiCudaRtENS_22AccGpuUniformCudaHipRtIS3_St17integral_constantImLm3EEjEES6_jN5cupla15cupla_cuda_sync11CuplaKernelIN8picongpu9particles9collision14InterCollisionILj64EEEEEJN5pmacc12ParticlesBoxINSH_5FrameINSH_15ParticlesBufferINSH_19ParticleDescriptionINSH_4meta6StringIJLc101EEEENSH_4math2CT6VectorIN4mpl_10integral_cIiLi8EEESU_NSS_2naEEEN5boost3mpl6v_itemINSB_13radiationMaskENSZ_INSB_13momentumPrev1ENSZ_INSB_9weightingENSZ_INSB_8momentumENSZ_INSB_8positionINSB_12position_picENSH_13pmacc_isAliasEEENSY_7vector0ISV_EELi0EEELi0EEELi0EEELi0EEELi0EEENSZ_INSB_11chargeRatioINSB_20ChargeRatioElectronsES16_EENSZ_INSB_9massRatioINSB_18MassRatioElectronsES16_EENSZ_INSB_7currentINSB_13currentSolver3EmZINSC_6shapes3PQSENS1M_8strategy16CachedSupercellsEEES16_EENSZ_INSB_13interpolationINSB_28FieldT
=========     by thread (32,0,0) in block (253,223,0)
=========     Address 0x7f2c543fff80 is out of bounds
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame:/lib/x86_64-linux-gnu/libcuda.so.1 [0x2e9441]
=========     Host Frame:/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/lib64/libcudart.so.11.0 [0xf62b]
=========     Host Frame:/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/lib64/libcudart.so.11.0 (cudaLaunchKernel + 0x1c1) [0x4f5b1]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x9cf8a]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x2cf8c2]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x2fca4f]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x3411e4]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x7c039]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x73399]
=========     Host Frame:/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xf3) [0x24083]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x74bae]
=========
[cupla] Error: </home/bastra54/src/picongpu/include/pmacc/../pmacc/eventSystem/events/kernelEvents.hpp>:165 Last error after kernel launch N8picongpu9particles9collision14InterCollisionILj64EEE [/home/bastra54/src/picongpu/include/pmacc/../picongpu/particles/collision/InterCollision.hpp:354 ]
[cupla] Error: </home/bastra54/src/picongpu/include/pmacc/../pmacc/eventSystem/events/kernelEvents.hpp>:165 Last error after kernel launch N8picongpu9particles9collision14InterCollisionILj64EEE [/home/bastra54/src/picongpu/include/pmacc/../picongpu/particles/collision/InterCollision.hpp:354 ]
========= Program hit cudaErrorLaunchFailure (error 719) due to "unspecified launch failure" on CUDA API call to cudaStreamSynchronize.
=========     Saved host backtrace up to driver entry point at error
=========     Host Frame:/lib/x86_64-linux-gnu/libcuda.so.1 [0x423ce6]
=========     Host Frame:/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/lib64/libcudart.so.11.0 (cudaStreamSynchronize + 0x144) [0x4f3b4]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x2cf6f8]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x2fca4f]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x3411e4]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x7c039]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x73399]
=========     Host Frame:/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xf3) [0x24083]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x74bae]
=========
========= Program hit cudaErrorLaunchFailure (error 719) due to "unspecified launch failure" on CUDA API call to cudaGetLastError.
=========     Saved host backtrace up to driver entry point at error
=========     Host Frame:/lib/x86_64-linux-gnu/libcuda.so.1 [0x423ce6]
=========     Host Frame:/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/lib64/libcudart.so.11.0 (cudaGetLastError + 0x11f) [0x3773f]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x2cf6fd]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x2fca4f]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x3411e4]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x7c039]
=========     Host Frame:/home/bastra54/picOutputsdebug_5/input/bin/picongpu [0x73399]
=========     Host Frame:/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xf3) [0x24083]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x74bae]
=========
========= Program hit cudaErrorLaunchFailure (error 719) due to "unspecified launch failure" on CUDA API call to cudaEventQuery.
=========     Saved host backtrace up to driver entry point at error
=========     Host Frame:/lib/x86_64-linux-gnu/libcuda.so.1 [0x423ce6]
=========     Host Frame:/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/lib64/libcudart.so.11.0 (cudaEventQuery + 0x133) [0x3b013]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x46ec8a]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x76219]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x77386]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x136e38]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x11f170]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0xac18d]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x69843]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x73399]
=========     Host Frame:/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xf3) [0x24083]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x74bae]
=========
========= Program hit cudaErrorLaunchFailure (error 719) due to "unspecified launch failure" on CUDA API call to cudaGetLastError.
=========     Saved host backtrace up to driver entry point at error
=========     Host Frame:/lib/x86_64-linux-gnu/libcuda.so.1 [0x423ce6]
=========     Host Frame:/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/lib64/libcudart.so.11.0 (cudaGetLastError + 0x11f) [0x3773f]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x15f5b1]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x46edff]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x76219]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x77386]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x136e38]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x11f170]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0xac18d]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x69843]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x73399]
=========     Host Frame:/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xf3) [0x24083]
=========     Host Frame:/home/bastra54/picOutputs/debug_5/input/bin/picongpu [0x74bae]
=========
terminate called after throwing an instance of 'std::runtime_error'
  what():  /home/bastra54/src/picongpu/thirdParty/cupla/alpaka/include/alpaka/event/EventUniformCudaHipRt.hpp(148) 'ret = TApi::eventQuery(event.getNativeHandle())' returned error  : 'cudaErrorLaunchFailure': 'unspecified launch failure'!
========= Error: process didn't terminate successfully
========= No CUDA-MEMCHECK results found

```

So it also shows a crash when accessing memory inside collisions, but again it is not clear which part to blame for it

**Comment by psychocoderHPC** [comment]
The crash happen during an atomic operation in inter collision functor. This already helps to to reduce the code we need to review again.

Unfortionally collision is using mallocMC, this mean the error could happen there too.

**Comment by sbastrakov** [comment]
`Address 0x7f2c543fff80 is out of bounds` - i think this address is definitely too large, so if mallocMC gave it it's wrong? Or perhaps we somehow have a bug that results in somehow improper address

**Comment by psychocoderHPC** [comment]
The crashing function `/opt/spack-modules/linux-ubuntu20.04-skylake_avx512/gcc-10.2.0/cuda-11.0.2-mtxiuzc35ykbmjk4fnpmhwnjbgquvcv5/targets/x86_64-linux/include/device_atomic_functions.hpp:157` points to

```C++
__DEVICE_ATOMIC_FUNCTIONS_DECL__ unsigned int atomicOr(unsigned int *address, unsigned int val)
{
  return __uAtomicOr(address, val);
}
```

**Comment by sbastrakov** [comment]
During investigation, together with @psychocoderHPC we found a [bug in mallocMC](https://github.com/alpaka-group/mallocMC/issues/236), same as also reported today in mattermost. This bug seems to exist in its code base for a long time, but appears unlikely to actually occur in a PIConGPU simulation - a few things have to go "right" for it. Which they did when I was running this setup on our dev system.

I am actually not sure if this is what you observe with this issue - there may be some other issue as well. I suggest waiting for the fix and then trying to run on a fresh version.

**Comment by Anton-Le** [comment]
Once the mallocMC problem is fixed I'll try with the original set-up. My guess is that the problem reported here is not, or only indirectly related to the mallocMC bug. As stated: we have not observed the NaN problem with 'old' versions (16.06.2022 and prior), so if the mallocMC bug existed at that time it's likely not the culprit.

**Comment by Anton-Le** [comment]
I have tried the same set-up with the newest PIConGPU `dev`, incorporating #4220 . The problem of `nan` field values persists with the mallocMC fix, too.

The following are the contents of `fields_energy.dat`:
```bash 
#step total[Joule] Bx[Joule] By[Joule] Bz[Joule] Ex[Joule] Ey[Joule] Ez[Joule] 
0 0.000000000000000e+00 0 0 0 0 0 0
1 nan nan nan 9.48154e-16 3.74698e-15 3.74375e-15 nan
```

**Comment by sbastrakov** [comment]
FYI today we also found out that in 2d simulations at some point we do incorrect access of `simDim`-dimensional vectors. So for `v` of type `pmacc::Vector<T, 2>` we somehow do out-of-bounds access `v[2]`. Investigating where and why it happens.

Edit: turned out harmless, see next message.

**Comment by sbastrakov** [comment]
Okay, that last thing turned out a quite harmless PNG stuff, fixed by #4225 .

**Comment by sbastrakov** [comment]
Fix is merged to `dev`, for my runs on Hemera the problem is no longer observed. Closing the issue. Please reopen/make new issues if something goes wrong again.
